servers/src/main/java/tachyon/worker/block/evictor/GreedyEvictor.java

```diff
@@ -0,0 +1,160 @@
+          availableBytes);
+      return null;
+    }
+
+    // 2. Check if the selected StorageDir already has enough space.
+    List<Pair<Long, BlockStoreLocation>> toTransfer =
+        new ArrayList<Pair<Long, BlockStoreLocation>>();
+    List<Long> toEvict = new ArrayList<Long>();
+    long bytesAvailableInDir = selectedDir.getAvailableBytes();
+    if (bytesAvailableInDir >= availableBytes) {
+      // No need to evict anything, return an eviction plan with empty instructions.
+      return new EvictionPlan(toTransfer, toEvict);
+    }
+
+    // 3. Collect victim blocks from the selected StorageDir. They could either be evicted or
+    // moved.
+    List<BlockMeta> victimBlocks = new ArrayList<BlockMeta>();
+    for (BlockMeta block : selectedDir.getBlocks()) {
+      victimBlocks.add(block);
+      bytesAvailableInDir += block.getBlockSize();
+      if (bytesAvailableInDir >= availableBytes) {
+        break;
+      }
+    }
+
+    // 4. Make best effort to transfer victim blocks to lower tiers rather than evict them.
+    Map<StorageDir, Long> pendingBytesInDir = new HashMap<StorageDir, Long>();
+    for (BlockMeta block : victimBlocks) {
+      StorageTier fromTier = block.getParentDir().getParentTier();
+      List<StorageTier> toTiers = mMetaManager.getTiersBelow(fromTier.getTierAlias());
+      StorageDir toDir = selectDirToTransferBlock(block, toTiers, pendingBytesInDir);
+      if (toDir == null) {
+        // Not possible to transfer
+        toEvict.add(block.getBlockId());
+      } else {
+        StorageTier toTier = toDir.getParentTier();
+        toTransfer.add(new Pair<Long, BlockStoreLocation>(block.getBlockId(),
+            new BlockStoreLocation(toTier.getTierAlias(), toTier.getTierLevel(), toDir
+                .getDirIndex())));
+        if (pendingBytesInDir.containsKey(toDir)) {
+          pendingBytesInDir.put(toDir, pendingBytesInDir.get(toDir) + block.getBlockSize());
+        } else {
+          pendingBytesInDir.put(toDir, block.getBlockSize());
+        }
+      }
+    }
+    return new EvictionPlan(toTransfer, toEvict);
+  }
+
+  private boolean canEvictBlocksFromDir(StorageDir dir, long availableBytes) {
+    return dir.getCapacityBytes() > availableBytes;
+  }
+
+  private StorageDir selectDirToEvictBlocksFromAnyTier(long availableBytes) {
+    for (StorageTier tier : mMetaManager.getTiers()) {
+      for (StorageDir dir : tier.getStorageDirs()) {
+        if (canEvictBlocksFromDir(dir, availableBytes)) {
+          return dir;
+        }
+      }
+    }
+    return null;
+  }
+
+  private StorageDir selectDirToEvictBlocksFromTier(StorageTier tier, long availableBytes) {
+    for (StorageDir dir : tier.getStorageDirs()) {
+      if (canEvictBlocksFromDir(dir, availableBytes)) {
+        return dir;
+      }
+    }
+    return null;
+  }
+
+  private StorageDir selectDirToTransferBlock(BlockMeta block, List<StorageTier> toTiers,
```

Comment:
@shimingfei , That is a good point.

I think currently, since eviction is synchronous, synchronous shuffling of too much data may be too expensive and slow. However, when async eviction is supported, I think more advanced and "smarter" algorithms will be feasible.